renter/hostdb/scan.go

```diff
@@ -0,0 +1,288 @@
+// every few hours to see who is online and available for uploading.
+func (db *DB) threadedScan() {
+	if err := db.tg.Add(); err != nil {
+		return
+	}
+	defer db.tg.Done()
+
+	// The initial scan might have been interrupted. Queue one scan for every
+	// announced host that was missed by the initial scan and wait for the
+	// scans to finish before starting the scan loop.
+	pks, err := db.Hosts()
+	if err != nil {
+		return
+	}
+	db.mu.Lock()
+	for _, pk := range pks {
+		host, err := db.Host(pk)
+		if err != nil {
+			return
+		}
+		if len(host.ScanHistory) == 0 && host.HistoricUptime == 0 && host.HistoricDowntime == 0 {
+			db.queueScan(pk)
+		}
+	}
+	db.mu.Unlock()
+
+	// Do nothing until the scan list is empty. If there are hosts in the scan
+	// list, other threads are ensuring they all get scanned.
+	db.waitForScans()
+
+	db.mu.Lock()
+	// Set the flag to indicate that the initial scan is complete.
+	db.initialScanComplete = true
+	db.mu.Unlock()
+
+	for {
+		// Set up a scan for the hostCheckupQuantity most valuable hosts in the
+		// hostdb. Hosts that fail their scans will be docked significantly,
+		// pushing them further back in the hierarchy, ensuring that for the
+		// most part only online hosts are getting scanned unless there are
+		// fewer than hostCheckupQuantity of them.
+
+		// Grab a set of hosts to scan, grab hosts that are active, inactive, offline
+		// and known to get high diversity.
+		var onlineHosts, offlineHosts, knownHosts []types.PublicKey
+		allHosts, err := db.Hosts()
+		if err != nil {
+			return
+		}
+		for i := len(allHosts) - 1; i >= 0; i-- {
+			if len(onlineHosts) >= hostCheckupQuantity &&
+				len(offlineHosts) >= hostCheckupQuantity {
+				break
+			}
+
+			// Figure out if the host is known, online or offline.
+			host, err := db.Host(allHosts[i])
+			if err != nil {
+				return
+			}
+
+			online := len(host.ScanHistory) > 0 && host.ScanHistory[len(host.ScanHistory)-1].Success
+			if online && len(onlineHosts) < hostCheckupQuantity {
+				onlineHosts = append(onlineHosts, allHosts[i])
+			} else if !online && len(offlineHosts) < hostCheckupQuantity {
+				offlineHosts = append(offlineHosts, allHosts[i])
+			}
+		}
+
+		// Queue the scans for each host.
+		db.mu.Lock()
+		for _, pk := range knownHosts {
+			db.queueScan(pk)
+		}
+		for _, pk := range onlineHosts {
+			db.queueScan(pk)
+		}
+		for _, pk := range offlineHosts {
+			db.queueScan(pk)
+		}
+		db.mu.Unlock()
+
+		// Sleep for a random amount of time before doing another round of
+		// scanning. The minimums and maximums keep the scan time reasonable,
+		// while the randomness prevents the scanning from always happening at
+		// the same time of day or week.
+		sleepRange := uint64(maxScanSleep - minScanSleep)
+		sleepTime := minScanSleep + time.Duration(frand.Uint64n(sleepRange))
+
+		// Sleep until it's time for the next scan cycle.
+		select {
+		case <-db.tg.StopChan():
+			return
+		case <-time.After(sleepTime):
+		}
+	}
```

Comment:
so complicated! Let's do something like this (pseudocode) instead:
```go

func (db *DB) scanLoop() {
	scanCh := make(chan struct{}, maxInflightScans)
	for i := 0; i < maxInflightScans; i++ {
		go func() {
			for {
				host := <-scanCh
				res := scanHost(host)
				db.ModifyHost(host, res)
			}
		}()
	}

	for {
		host, err := store.HostForScan(since)
		if err == NoUnscannedHosts {
			// sleep for a while
			time.Sleep(aWhile)
			continue
		}
		scanCh <- host
	}
}
```